pkg/endpoint/public_ip_watcher.go

```diff
@@ -0,0 +1,183 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoint
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/log"
+	"github.com/submariner-io/admiral/pkg/watcher"
+	submv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	v1 "github.com/submariner-io/submariner/pkg/client/clientset/versioned/typed/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+)
+
+type Config struct {
+	WatcherConfig *watcher.Config
+	SubmSpec      *types.SubmarinerSpecification
+	Interval      time.Duration
+}
+
+type PublicIPWatcher struct {
+	mutex           sync.Mutex
+	k8sClient       kubernetes.Interface
+	endpoints       v1.EndpointInterface
+	endpointWatcher watcher.Interface
+	localEndpoint   *submv1.Endpoint
+	config          *Config
+}
+
+const DefaultMonitorInterval = 20 * time.Second
+
+func NewPublicIPWatcher(config *Config, k8sClient kubernetes.Interface, endpoints v1.EndpointInterface) (*PublicIPWatcher, error) {
+	controller := &PublicIPWatcher{
+		config:    config,
+		k8sClient: k8sClient,
+		endpoints: endpoints,
+	}
+
+	if config.Interval == 0 {
+		config.Interval = DefaultMonitorInterval
+	}
+
+	config.WatcherConfig.ResourceConfigs = []watcher.ResourceConfig{
+		{
+			Name:         "Endpoint Controller for Monitoring Public IP",
+			ResourceType: &submv1.Endpoint{},
+			Handler: watcher.EventHandlerFuncs{
+				OnCreateFunc: controller.endpointCreatedorUpdated,
+				OnUpdateFunc: controller.endpointCreatedorUpdated,
+				OnDeleteFunc: controller.endpointDeleted,
+			},
+			SourceNamespace: config.SubmSpec.Namespace,
+		},
+	}
+
+	var err error
+
+	controller.endpointWatcher, err = watcher.New(config.WatcherConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating endpoint watcher")
+	}
```

Comment:
We could pass in the local `Endpoint` as we do for other gateway components since its identifying information doesn't change for the lifetime of the gateway process. You can use `util.GetEndpointCRDName` when updating as the datastore syncer does. This would avoid the added overhead of watching and simplify the code.

Reply:
Sure, will make the necessary changes. Thanks for reviewing @tpantelis